### ISSUE TITLE
Refactor Geometry to be physics-agnostic

### ISSUE THREAD

**Issue by jvmadan** [climate-analytics-lab/jax-gcm#230]
our Geometry struct should be suited for any possible Physics module, at the moment it's lifted directly from Speedy. It might not take much refactoring to clean this up, since the fields in Geometry are mostly things that are generally useful for any physics module anyway, but at the very least we should change it so the sigma_layer_boundaries dictionary is provided by the physics module.

**Comment by jvmadan** [comment]
* Geometry is serving two different purposes right now--it's 1) our user interface for the model grid, and it's also 2) the interface we pass to the physics tendency methods
* The main reason not to just use dinosaur coordinatesystem for purpose 1 is that we want this user interface to also handle the terrain file read-in - my intuition is that we should keep that out of the model constructor, to match the ForcingData pattern and I think just a vague sense that our model constructor used to take more arguments and I remember that feeling worse
* I think there are changes to be made on the PhysicsData side / arguments to physics tendencies side of the equation though.
* Since the Physics module is known at model construction, we could have physics modules provide their own GeometryData struct (since we had to take the physics-module-specific fields out of Geometry anyway) which would take the model coords (or whatever) and compute the relevant fields, saving the GeometryData as a Model or Physics field. Am I forgetting a reason that we aren't already doing this to avoid re-computing fields that are the same every timestep (at least the ones that don't depend on forcings, tyear, etc? We could even name it StaticData, throw static parameters like `lfluxland` in there, mark it as a static argument on the physics tendency functions, and then get to use regular if statements rather than jax.lax.cond in surface fluxes, etc

**Comment by eldavenport** [comment]
Having terrain data separate also simplifies some of the invisible dependencies between lfluxland, forcingdata, and geometry (geometry has fmask, but lfluxland is in forcing data but we only want to set lfluxland if we have terrain which is read in by geometry). In this case lfluxland could just go in terrain data along with fmask and other things.

**Comment by duncanwp** [comment]
Adding @shoyer for visibility in case you have any strong preferences for dealing with land masks, orography and things like layer thickness.

**Comment by eldavenport** [comment]
> OK, I think this makes sense:
> 
> **Proposed Refactoring: Geometry → TerrainData + Physics-internal Coord Caching**

I like this proposed refactor, that’s pretty much what I was envisioning. I’m sure it might change a little, but I am in favor of going forward with this. As for the name, I’m ok with TerrainData or OrographyData ? No strong preferences.
